pkg/netbox/api/ip_range_claim.go

```diff
@@ -0,0 +1,190 @@
+)
+
+func (r *NetboxClient) RestoreExistingIpRangeByHash(hash string) (*models.IpRange, error) {
+	customIpRangeSearch := newQueryFilterOperation(nil, []CustomFieldEntry{
+		{
+			key:   config.GetOperatorConfig().NetboxRestorationHashFieldName,
+			value: hash,
+		},
+	})
+	list, err := r.Ipam.IpamIPRangesList(ipam.NewIpamIPRangesListParams(), nil, customIpRangeSearch)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: find a better way?
+	if list.Payload.Count != nil && *list.Payload.Count == 0 {
+		return nil, nil
+	}
+
+	// We should not have more than 1 result...
+	if len(list.Payload.Results) != 1 {
+		return nil, fmt.Errorf("incorrect number of restoration results, number of results: %v", len(list.Payload.Results))
+	}
+	res := list.Payload.Results[0]
+	if res.StartAddress == nil || res.EndAddress == nil {
+		return nil, errors.New("invalid IP range")
+	}
+
+	return &models.IpRange{
+		StartAddress: *res.StartAddress,
+		EndAddress:   *res.EndAddress,
+		Id:           res.ID,
+	}, nil
+}
+
+// GetAvailableIpRangeByClaim searches an available IpRange in Netbox matching IpRangeClaim requirements
+func (r *NetboxClient) GetAvailableIpRangeByClaim(ipRangeClaim *models.IpRangeClaim) (*models.IpRange, error) {
+	_, err := r.GetTenantDetails(ipRangeClaim.Metadata.Tenant)
+	if err != nil {
+		return nil, err
+	}
+
+	responseParentPrefix, err := r.GetPrefix(&models.Prefix{
+		Prefix:   ipRangeClaim.ParentPrefix,
+		Metadata: ipRangeClaim.Metadata,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(responseParentPrefix.Payload.Results) == 0 {
+		return nil, utils.NetboxNotFoundError("parent prefix")
+	}
+
+	parentPrefixId := responseParentPrefix.Payload.Results[0].ID
+	responseAvailableIPs, err := r.GetAvailableIpAddressesByParentPrefix(parentPrefixId)
+	if err != nil {
+		return nil, err
+	}
+
+	startAddress, endAddress, err := searchAvailableIpRange(responseAvailableIPs, ipRangeClaim.Size)
+	if err != nil {
+		return nil, err
+	}
+
+	startAddress, err = r.SetIpAddressMask(startAddress, responseAvailableIPs.Payload[0].Family)
+	if err != nil {
+		return nil, err
+	}
+
+	endAddress, err = r.SetIpAddressMask(endAddress, responseAvailableIPs.Payload[0].Family)
+	if err != nil {
+		return nil, err
+	}
+
+	return &models.IpRange{
+		StartAddress: startAddress,
+		EndAddress:   endAddress,
+	}, nil
+}
+
+// GetAvailableIpsByIpRange returns all available Ips in Netbox matching IpRangeClaim requirements
+func (r *NetboxClient) GetAvailableIpAddressesByIpRange(ipRangeId int64) (*ipam.IpamIPRangesAvailableIpsListOK, error) {
+	requestAvailableIPs := ipam.NewIpamIPRangesAvailableIpsListParams().WithID(ipRangeId)
+	responseAvailableIPs, err := r.Ipam.IpamIPRangesAvailableIpsList(requestAvailableIPs, nil)
+	if err != nil {
+		return nil, err
+	}
+	return responseAvailableIPs, nil
+}
+
+func searchAvailableIpRange(availableIps *ipam.IpamPrefixesAvailableIpsListOK, requiredSize int) (string, string, error) {
```

Comment:
Just curious, if `availableIps` is guaranteed to be sorted.

Reply:
The list returned when calling the /api/ipam/prefixes/<id>/available-ips/ is sorted. 
But I now also added sorting to this function in case this would change in future.

Reply:
Cool, thanks!